k path for the original cell without standardization

jaemolihm opened the pull request:
As I discussed with @giovannipizzi during the Wannier 2022 developers meeting, I added two functions `get_path_orig_cell` and `get_explicit_k_path_orig_cell`, which does the same thing with the existing functions except that they keep the original cell: the original cell is not standardized or symmetrized.

The idea for the implementation is that linear combination of the lattice vectors does not change the 1st BZ, so only the real space rotations needs to be taken into account. I first implemented this idea in `Brillouin.jl`, a Julia implementation of the seekpath algorithm. For further information, see discussions in https://github.com/thchr/Brillouin.jl/pull/15, https://github.com/thchr/Brillouin.jl/pull/16.

If the input cell is a non-standard primitive unit cell, the returned k path is equivalent to the k path for the standard cell.
If the input cell is a supercell of a smaller primitive cell, the returned k path is that of the associated primitive cell, in the basis of supercell reciprocal lattice.

### Example: silicon
* Non-standard unit cell is rotated by 17 degrees and the basis vectors are linearly combined.
* The supercell is a 1\*1\*2 supercell of the non-standard primitive cell.
* Standard and non-standard unit cell show identical band structures.
* Supercell band structure includes an identical copy of the primitive cell band structure and additional folded bands.
* Input files available in https://github.com/jaemolihm/seekpath_orig_cell_example
![image](https://user-images.githubusercontent.com/32537613/171751563-1ae0527b-2f8e-4af2-9e4d-a80803bd89c2.png)

### Tests
* `TestPaths3D_HPKOT_Orig_Cell.base_test`: Test the k path with and without standardization are equivalent up to rotation.
* `TestPaths3D_HPKOT_Orig_Cell.test_no_symmetrization`: Check that `get_path_orig_cell` does not symmetrize the input structure so that for a slightly distorted system the k points are not exactly on the high-symmetry points
* `TestExplicitPaths_Orig_Cell.test_keys`: Test output keys of `get_explicit_k_path_orig_cell`
* `TestExplicitPaths_Orig_Cell.test_path`: Test the explicit k point list for with and without standardization are equivalent up to rotation.

jaemolihm commented:
This PR does not "Provide Transformation Matrix as output of the code" which is the title of issue #16, but it solves the related issue of calculating the k path for non-standard cells (#77, https://github.com/aiidateam/aiida-core/issues/4391).

tkotani commented:
This is what I implemented several years ago.
As you noticed, we have to find out unimodular matrix which connects two sets of primitive vectors.
FYI: Related to "lattice encryption", we don't have very efficient way to find out the unimodular matrix.
Some trial and error procedure needed, I think. (this is what I had implemented).
Math of integer matrix is not so easy to handle.

jaemolihm commented:
> This is what I implemented several years ago. As you noticed, we have to find out unimodular matrix which connects two sets of primitive vectors. FYI: Related to "lattice encryption", we don't have very efficient way to find out the unimodular matrix. Some trial and error procedure needed, I think. (this is what I had implemented). Math of integer matrix is not so easy to handle.

Dear @tkotani , thanks for the comments. To be clear, I did not calculate or use the integer-valued unimodular matrix. Instead, I used the orthogonal rotation matrix that describes the real-space rotation (`std_rotation_matrix` of spglib).

Here, I did not need the integer-valued matrix. If in the future one wants it for some reason, one can use L_std = R * L_inp * N (L_std: standard primitive lattice, L_inp: input lattice, R: orthogonal rotation matrix, N: integer-valued unimodular matrix). Since L_std, R, L_inp are known, N can be easily computed.

tkotani commented:
Dear @jaemolihm, Thank you for your reply. I see.
Generally speaking, the same lattice is defined by different sets of primitive vectors.
The sets are related by unimodular matrix. 
 
See https://cseweb.ucsd.edu/classes/wi12/cse206A-a/lec1.pdf
Theorem 4. (page 2.) is what I said. We have to find out the matrix U.

giovannipizzi commented on docs/source/maindoc.rst:
```diff
@@ -61,24 +61,42 @@ The output of the function is a dictionary containing, among other quantities, t
+If the input cell is a supercell of a smaller primitive cell, the returned k path is that of the associated primitive cell, in the basis of supercell reciprocal lattice.
+In this case, the k points are **not** the high-symmetry points of the first Brillouin zone of the given supercell, but the high-symmetry points of the Brillouin zone of the associated primitive cell.
+
+Note that contrary to ``get_path``, ``get_path_orig_cell`` calculates the k path based on the symmetrized structure but does not symmetrize the input structure itself.
+Hence, if the symmetry of the input structure is slightly broken below the symmetry precision ``symprec``, the output k points may not be exactly on the high-symmetry k points.
```
```suggestion
**Note that contrary to ``get_path``, ``get_path_orig_cell`` calculates the k path based on the symmetrized structure but does not symmetrize the input structure itself.**
Hence, if the symmetry of the input structure is slightly broken below the symmetry precision ``symprec``, the **output k points may not be exactly on the high-symmetry k points.**
```

giovannipizzi commented on docs/source/maindoc.rst:
```diff
@@ -61,24 +61,42 @@ The output of the function is a dictionary containing, among other quantities, t
-then need to figure out how to remap the labeled k-points in the choice of
-cell you did.
+If you want a k-point path for a non-standardized cell, you can use the
+``get_path_orig_cell`` and ``get_explicit_k_path_orig_cell`` functions: see the above subsection.
```
```suggestion
``get_path_orig_cell`` and ``get_explicit_k_path_orig_cell`` functions: see the above subsection (and in particular, check the limitations if the symmetry of the input structure is slightly broken).
```

giovannipizzi commented on seekpath/getpaths.py:
```diff
@@ -325,3 +325,268 @@ def get_explicit_k_path(
+    r"""
+    Return the kpoint path information for band structure given a
+    crystal structure, using the paths from the chosen recipe/reference.
+    The original unit cell is used. Standardization or symmetrization of the
+    input unit cell is not performed.
```
```suggestion
    The original unit cell (i.e., the one provided in input by the user) is used. 
    Standardization or symmetrization of the input unit cell is not performed.
```

giovannipizzi commented on seekpath/getpaths.py:
```diff
@@ -325,3 +325,268 @@ def get_explicit_k_path(
+
+    If the provided unit cell is a supercell of a smaller primitive cell,
+    return the standard k path of the smaller primitive cell in the basis
+    of the supercell reciprocal lattice vectors. In this case, the k point
+    labels lose their meaning: they are not at the high-symmetry points of
+    the first BZ of the given supercell.
```
```suggestion
    of the supercell reciprocal lattice vectors. Note that, in this case,
    the k-point labels are not at the high-symmetry points of
    the first BZ of the given supercell.
```

jaemolihm replied:
I did a slightly modified version: "In this case, the k-point labels lose their meaning as the corresponding k-points are not at the high-symmetry"

giovannipizzi commented on seekpath/getpaths.py:
```diff
@@ -325,3 +325,268 @@ def get_explicit_k_path(
+    is_supercell = abs(res["volume_original_wrt_prim"] - 1) > 0.1
+
+    if is_supercell:
+        print(
```
Replace with a warning instead, so it can be easily filtered

jaemolihm replied:
Done

giovannipizzi commented on seekpath/hpkot/__init__.py:
```diff
@@ -120,6 +120,8 @@ def get_path(
         - ``inverse_primitive_transformation_matrix``: the inverse of the matrix :math:`P`
           (the determinant is integer and gives the ratio in volume between
           the conventional and primitive cells)
+        - ``rotation_matrix``: rotation matrix in the Cartesian space from the input
```
```suggestion
        - ``rotation_matrix``: rotation matrix in Cartesian space from the input
```

jaemolihm commented:
I locally tested with spglib-2.0.2 and tests passed.

giovannipizzi reviewed (approved):
Thanks a lot, looks great!

jaemolihm commented:
Thanks a lot for the review!
